### PR TITLE
Fix/check hidden character

### PR DESCRIPTION
**Pull request by luciecastella**
First half of #164 resolution. 

The error message in case of special character is: 
ValueError: Unexpected special character(s) in: scenarios.yaml, line 3, col 12. variables.yaml, line 5, col 53.

Tell me if it could be written in a more understandable way.

### CODE REVIEW

File: tests/test_testing.py [hunk 1]
```diff
@@ -26,3 +26,10 @@ def test_assert_yaml_fails(caplog):
     obs = log[2].replace("\n", "")  # strip newlines from observed log message
     exp = r"Error parsing file while scanning a simple key.*, line 4, column 1"
     assert re.match(exp, obs)
+
+
+def test_special_character():
```
Comment (phackstock) [inline tests/test_testing.py:31]: I would call this "hidden character".
```suggestion
def test_hidden_character():
```

File: tests/test_testing.py [hunk 2]
```diff
@@ -26,3 +26,10 @@ def test_assert_yaml_fails(caplog):
...
+    """Check that a non-printable character in any yaml file will raise an error"""
+    match = "variables.yaml, line 5, col 53."
+    with pytest.raises(ValueError, match=match):
+        assert_valid_yaml(TEST_DATA_DIR / "special_character")
```
Comment (phackstock) [inline tests/test_testing.py:35]: I would also rename the folder where the test data are located
```suggestion
        assert_valid_yaml(TEST_DATA_DIR / "hidden_character")
```

File: tests/test_testing.py [hunk 3]
```diff
@@ -26,3 +26,10 @@ def test_assert_yaml_fails(caplog):
     obs = log[2].replace("\n", "")  # strip newlines from observed log message
     exp = r"Error parsing file while scanning a simple key.*, line 4, column 1"
     assert re.match(exp, obs)
+
+
+def test_special_character():
+    """Check that a non-printable character in any yaml file will raise an error"""
+    match = "variables.yaml, line 5, col 53."
```
Comment (phackstock) [inline tests/test_testing.py:33]: If you only test for the invisible character in variables.yaml, which totally fine, you should leave `data/special_character/definitions/scenario/scenarios.yaml` out as it is not used in the test.

File: tests/data/special_character/definitions/variable/tag_fuel.yaml [hunk 4]
```diff
@@ -0,0 +1,3 @@
+- Fuel:
```
Comment (phackstock) [inline tests/data/special_character/definitions/variable/tag_fuel.yaml:1]: Don't think you need the tag codelist for testing for hidden characters.

File: tests/data/special_character/definitions/scenario/scenarios.yaml [hunk 5]
```diff
@@ -0,0 +1,4 @@
+- Scenario1
```
Comment (phackstock) [inline tests/data/special_character/definitions/scenario/scenarios.yaml:1]: If this codelist is not used in the test you can remove this file altogether.

File: tests/data/special_character/definitions/variable/variables.yaml [hunk 6]
```diff
@@ -0,0 +1,9 @@
+- Primary Energy:
```
Comment (phackstock) [inline tests/data/special_character/definitions/variable/variables.yaml:1]: I would shorten this file to include maybe only the variable with the hidden character. Also I don't think the tags are needed.

**Review by phackstock — CHANGES_REQUESTED** [review]
Solid start to the PR, some comments below.

**Review by phackstock — APPROVED** [review]
Looks all good to me.

File: tests/data/hidden_character/definitions/scenario/scenarios.yaml [hunk 7]
```diff
@@ -0,0 +1,4 @@
+- Scenario1
+- Scenario2 
+- Scenario3 
+- Scenario4
```
Comment (danielhuppmann) [inline tests/data/hidden_character/definitions/scenario/scenarios.yaml:4]: Please add an empty line at end of file

File: nomenclature/testing.py [hunk 8]
```diff
@@ -21,6 +24,18 @@ def assert_valid_yaml(path: Path):
             error = True
             logger.error(f"Error parsing file {e}")
 
+        with open(file, "r", encoding="utf-8") as all_lines:
+            # check if any special character is found in the file
+            for index, line in enumerate(all_lines.readlines()):
+                for col, char in enumerate(line):
+                    if char not in printable:
+                        special_characters += (
+                            f"{file.name}, line {index + 1}, col {col + 1}. "
```
Comment (danielhuppmann) [inline nomenclature/testing.py:33]: Print error message with pointers to invalid characters as list.
```suggestion
                            f"\n - {file.name}, line {index + 1}, col {col + 1}. "
```

**Review by danielhuppmann — APPROVED** [review]
Two minor suggestions, then good to be merged!

**Pull request merged**